File: leetcode/21-mergeTwoLinkedLists/21.ts


class ListNode {
    val: number
    next: ListNode | null
    constructor(val?: number, next?: ListNode | null) {
        this.val = (val === undefined ? 0 : val)
        this.next = (next === undefined ? null : next)
    }
}


function mergeTwoLists(list1: ListNode, list2: ListNode): ListNode | null {
    !(list1?.val && list2?.val) ? null : [];
    var arr: number[] = [];
    const addToArray = (list: ListNode) => {
        list?.val ? arr.push(list.val) : null;
        if (list?.next == null) {
           return;
        } else {
            addToArray(list.next);
        }
    }

    addToArray(list1);
    addToArray(list2);

    return arr.sort((a, b) => a - b).map((val) => new ListNode(val)).map((node, index, a) => {
        node.next = a[index + 1] || null
        return node
    })[0] ?? null;
};

var list1 = [].map((val) => new ListNode(val ?? null)).map((node, index, arr) => {
    node.next = arr[index + 1] || null
    return node
})
var list2 = [].map((val) => new ListNode(val ?? null)).map((node, index, arr) => {
    node.next = arr[index + 1] || null
    return node
})

console.log(mergeTwoLists(list1[0], list2[0]))

File: leetcode/21-mergeTwoLinkedLists/21.go
package mergeTwoLists

import (
	"fmt"
)

/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */

type ListNode struct {
	Val  int
	Next *ListNode
}

func mergeTwoLists(list1 *ListNode, list2 *ListNode) *ListNode {
	var head, tail *ListNode
	for list1 != nil && list2 != nil {
		var node *ListNode
		if list1.Val < list2.Val {
			node = list1
			list1 = list1.Next
		} else {
			node = list2
			list2 = list2.Next
		}
		if head == nil {
			head = node
		} else {
			tail.Next = node
		}
		tail = node
	}
	if list1 != nil {
		if head == nil {
			head = list1
		} else {
			tail.Next = list1
		}
	}
	if list2 != nil {
		if head == nil {
			head = list2
		} else {
			tail.Next = list2
		}
	}
	return head
}
